src/compound_merkle.rs

```diff
@@ -0,0 +1,231 @@
+use crate::hash::Algorithm;
+use crate::merkle::{get_merkle_tree_len, Element, MerkleTree};
+use crate::proof::Proof;
+use crate::store::{Store, StoreConfig};
+use typenum::marker_traits::Unsigned;
+
+#[derive(Debug, Clone, Eq, PartialEq, Default)]
+pub struct CompoundMerkleTree<T, A, K, B, N>
+where
+    T: Element,
+    A: Algorithm<T>,
+    K: Store<T>,
+    B: Unsigned, // Branching factor of sub-trees
+    N: Unsigned, // Number of nodes at top layer
+{
+    trees: Vec<MerkleTree<T, A, K, B>>,
+    top_layer_nodes: usize,
+    len: usize,
+    leafs: usize,
+    height: usize,
+    root: T,
+
+    _n: PhantomData<N>,
+}
+
+impl<T: Element, A: Algorithm<T>, K: Store<T>, B: Unsigned, N: Unsigned>
+    CompoundMerkleTree<T, A, K, B, N>
+{
+    /// Creates new compound merkle tree from a vector of merkle trees.
+    pub fn new(trees: Vec<MerkleTree<T, A, K, B>>) -> Result<CompoundMerkleTree<T, A, K, B, N>> {
+        let top_layer_nodes = <N as Unsigned>::to_usize();
+        ensure!(
+            trees.len() == top_layer_nodes,
+            "Length of trees MUST equal the number of top layer nodes"
+        );
+        ensure!(
+            trees.iter().all(|ref mt| mt.height() == trees[0].height()),
+            "All passed in trees must have the same height"
+        );
+        ensure!(
+            trees.iter().all(|ref mt| mt.len() == trees[0].len()),
+            "All passed in trees must have the same length"
+        );
+
+        // Total number of leafs in the compound tree is the combined leafs total of all subtrees.
+        let leafs = trees.iter().fold(0, |leafs, mt| leafs + mt.leafs());
+        // Total length of the compound tree is the combined length of all subtrees plus the root.
+        let len = trees.iter().fold(0, |len, mt| len + mt.len()) + 1;
+        // Total height of the compound tree is the height of any of the sub-trees to top-layer plus root.
+        let height = trees[0].height() + 1;
+        // Calculate the compound root by hashing the top layer roots together.
+        let roots: Vec<T> = trees.iter().map(|x| x.root()).collect();
+        let root = A::default().multi_node(&roots, 1);
+
+        Ok(CompoundMerkleTree {
+            trees,
+            top_layer_nodes,
+            len,
+            leafs,
+            height,
+            root,
+            _n: PhantomData,
+        })
+    }
+
+    /// Given an already constructed merkle tree contained as a slice,
+    /// add it to this compound merkle tree.
+    pub fn new_from_slice(
+        tree_data: &[u8],
+        leafs: usize,
+    ) -> Result<CompoundMerkleTree<T, A, K, B, N>> {
+        let mut trees = Vec::with_capacity(1);
+        trees.push(MerkleTree::<T, A, K, B>::from_tree_slice(tree_data, leafs)?);
+
+        CompoundMerkleTree::new(trees)
+    }
+
+    /// Given an already constructed merkle tree contained as a slice,
+    /// add it to this compound merkle tree.
+    pub fn new_from_slice_with_config(
```

Comment:
this should be called `add_from_slice_with_config`

Reply:
sorry I got confused from the docs, I guess this just creates a new compound tree, not adds one, naming is good